_posts/2018-02-15-continuous-deployment-concourse.md

```diff
@@ -0,0 +1,286 @@
+- *Tasks* are like functions, with inputs and outputs. Since it's a container you can choose which image it will use to fit your needs.
+
+The *pipeline* contains all the jobs necessary to fulfill a purpose. When a resource gets updated it will trigger a *build* which is the logical execution of the job's tasks.
+
+## Build images
+
+### Docker image resource definition:
+```
+- name: foobar-app-image
+  type: docker-image
+  source:
+    repository: gcr.io/foobar/foobar_app
+    username: _json_key
+    password: {{gcloud_service_key}}
+```
+
+### Build and push the image:
+```
+- put: foobar-app-image
+  params:
+    build: artifact-app/
+    cache: true
+    cache_tag: latest
+    tag: metadata/short-commit-hash
+    tag_as_latest: true
+```
+Note: 
+- the artifact-app/ volume contains a Dockerfile and dependencies to build the image 
+- the `latest` tag will be cached on disk, if we have a lot of builds this can save time!
+- The metadata is a simple task that uses a `git` Docker image to retrieve the git short commit hash:
+    ```
+    ---
+    platform: linux
+
+    image_resource:
+      type: docker-image
+      source: {repository: greenwall/git}
+
+    params:
+      TERM: xterm-256color
+
+    inputs:
+      - name: code-base
+
+    outputs:
+      - name: metadata
+
+    run:
+      path: /bin/sh
+      args:
+      - -exc
+      - |
+        GIT_SHORT_HASH=`cd code-base && git log --pretty=format:'%h' -n 1`
+        echo $GIT_SHORT_HASH > metadata/short-commit-hash
+    ```
+
+## Run unit tests on pre-built container
+We could easily reuse that image for unit testing, tests that should not depend on anything else than the code itself. Here's an example on how to do it with php-unit:
+
+```
+- task: php-unit-tests
+  config:
+    platform: linux
+    image_resource:
+      type: docker-image
+      source:
+        repository: gcr.io/foobar/foobar_app
+        username: _json_key
+        password: {{gcloud_service_key}}
+    run:
+      path: sh
+      args:
+      - -exc
+      - |
+        /var/www/vendor/phpunit/phpunit/phpunit --testsuite=unit --bootstrap /var/www/includes/tests/bootstrap.php --configuration /var/www/includes/tests/phpunit-conf.xml
+  on_failure:
+    put: git-pr
+    params:
+      context: php-unit-tests
+      path: git-pr
+      status: failure
+- put: git-pr
+  params:
+    context: php-unit-tests
+    path: git-pr
+    status: success
+```
+
+
+## Continous deployment on staging/production
+### Tooling
+To manage releases on Kubernetes we use [Helm](http://helm.sh). This tool also require kubectl to be properly configured and since we are using GKE we need the [gcloud](https://cloud.google.com/sdk/downloads) cli. These tools do not exist on Concourse, they are too specific, this is why I built an [alpine-based image containing all these tools](https://github.com/baptiste-bonnaudet/gcloud-kubectl-helm).
```

Comment:
- `This tool also require kubectl to be properly`
  - `require` => `requires`
  - kubectl should be wrapped in backticks as it refers to a binary
- `configured and since we are using GKE`
  - use the Oxford comma
  - Acronyms should not be used unless their full name equivalents have been mentioned : `we are using the Google Kubernetes Engine (GKE)`